### PR TITLE
APF e2e test: wait for steady state before proceeding

### PR DESCRIPTION
**Pull request by adtac**
**What type of PR is this?**

/kind failing-test

**What this PR does / why we need it**: Moments after creation, the APF objects could still be unprocessed by the API server. We should wait for steady state before proceeding with the rest of the test. Builds on top of https://github.com/kubernetes/kubernetes/pull/96966 (will rebase after that PR is merged).

**Which issue(s) this PR fixes**:

Related to #96803 and https://github.com/kubernetes/kubernetes/issues/96893
(we'll need to observe test-grid for a few days before closing the flake issue)

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

cc @MikeSpreitzer @lavalamp @yue9944882 @tkashem @ahg-g 
/sig api-machinery

### CODE REVIEW

File: test/e2e/apimachinery/flowcontrol.go
Comment (MikeSpreitzer) [inline test/e2e/apimachinery/flowcontrol.go:357]: There is an inversion of logic here.  The current code returns `false` if the desired condition is present.  That is backwards.  The code should return `false` if the desired condition is _not_ present.

File: test/e2e/apimachinery/flowcontrol.go
Comment (MikeSpreitzer) [inline test/e2e/apimachinery/flowcontrol.go:381]: This is not the time series to look for, because this time series is not created until some requests have been classified to this priority level and begun executing.  Instead look at the corresponding time series of the `apiserver_flowcontrol_request_concurrency_limit` metric, that one gets created when the config consuming controller processes the PriorityLevelConfiguration object.

**Review by MikeSpreitzer — CHANGES_REQUESTED** [review]

File: test/e2e/apimachinery/flowcontrol.go
Comment (tkashem) [inline test/e2e/apimachinery/flowcontrol.go:358]: and when we find the `Dangling` condition, do we also want to check that its `status` is set to `False`?

File: test/e2e/apimachinery/flowcontrol.go
Comment (tkashem) [inline test/e2e/apimachinery/flowcontrol.go:364]: `getPriorityLevelConcurrency` function already gets the concurrency for a priority level, can we reuse it?

File: test/e2e/apimachinery/flowcontrol.go
Comment (MikeSpreitzer) [reply test/e2e/apimachinery/flowcontrol.go:358]: That would be a nice add, yes.

File: test/e2e/apimachinery/flowcontrol.go
Comment (MikeSpreitzer) [reply test/e2e/apimachinery/flowcontrol.go:364]: The problem with that func, in the current version of this PR, is that when the sought time series is _not_ found the test fails rather than waits and tries again.

File: test/e2e/apimachinery/flowcontrol.go
Comment (yue9944882) [inline test/e2e/apimachinery/flowcontrol.go:375]: > return false, nil

returning an io.EOF which is non-nil error will fail the test

File: test/e2e/apimachinery/flowcontrol.go
Comment (yue9944882) [reply test/e2e/apimachinery/flowcontrol.go:381]: @MikeSpreitzer but that doesn't reflect the connection between the flowschema and prioritylevel, by the shared concurrency we don't know whether they're bonded.

File: test/e2e/apimachinery/flowcontrol.go
Comment (MikeSpreitzer) [inline test/e2e/apimachinery/flowcontrol.go:80]: I still do not think this test should insist on particular UIDs in the case of a request that does not match the FlowSchema that this test creates.

File: test/e2e/apimachinery/flowcontrol.go
Comment (MikeSpreitzer) [reply test/e2e/apimachinery/flowcontrol.go:381]: If this func waits until the FlowSchema has been processed and the PriorityLevelConfiguration has been processed then it is guaranteed (by the current implementation) that the two are connected.  We could remove that reliance on that implementation detail from this test by making it wait until the FlowSchemaStatus says the dandling condition's value is false.

File: test/e2e/apimachinery/flowcontrol.go
Comment (tkashem) [reply test/e2e/apimachinery/flowcontrol.go:80]: @MikeSpreitzer this PR builds on top of https://github.com/kubernetes/kubernetes/pull/96966. I have already addressed this feedback in 96966. 

@adtac I think this PR can merge independently of https://github.com/kubernetes/kubernetes/pull/96966. You can copy `getFlowSchema` and `getPriorityLevelConfig` function in this PR. I am no longer using them in https://github.com/kubernetes/kubernetes/pull/96966

File: test/e2e/apimachinery/flowcontrol.go
Comment (adtac) [reply test/e2e/apimachinery/flowcontrol.go]: Done. I was certain it was the other way around, but I think I get the full picture now.

File: test/e2e/apimachinery/flowcontrol.go
Comment (adtac) [reply test/e2e/apimachinery/flowcontrol.go]: Done: re-used the getPLConcurrency and also added a check to make sure the condition is true.

File: test/e2e/apimachinery/flowcontrol.go
Comment (adtac) [reply test/e2e/apimachinery/flowcontrol.go:361]: Done.

File: test/e2e/apimachinery/flowcontrol.go
Comment (adtac) [reply test/e2e/apimachinery/flowcontrol.go]: refactored things around slightly to return an error instead so that we can re-use this

File: test/e2e/apimachinery/flowcontrol.go
Comment (adtac) [reply test/e2e/apimachinery/flowcontrol.go]: Done, removed the cherry-picked commit and moved getFlowSchema into this commit

File: test/e2e/apimachinery/flowcontrol.go
Comment (MikeSpreitzer) [reply test/e2e/apimachinery/flowcontrol.go:358]: When the dangling Condition exists with the value "true" this means the apiserver has processed the FlowSchema object and _not_ found the corresponding PriorityLevelConfiguration object.  In this case this func should `return false, nil` to cause a pause and retry.

When the dangling Condition exists with a value of "true", this func should `return true, nil` to cause successful completion of the `wait.Poll`.

**Comment by MikeSpreitzer** [comment]
#97002

**Comment by MikeSpreitzer** [comment]
The logic in the lambda passed to `wait.Poll` inside `waitForSteadyState` still has a bit flipped.

File: test/e2e/apimachinery/flowcontrol.go
Comment (yue9944882) [inline test/e2e/apimachinery/flowcontrol.go:353]:
```suggestion
		if condition == nil || condition.Status != flowcontrol.ConditionFalse {
```

flipping the logic will solve the test failure

File: test/e2e/apimachinery/flowcontrol.go
Comment (MikeSpreitzer) [inline test/e2e/apimachinery/flowcontrol.go:356]: This comment still needs to be fixed.

Comment (adtac) [reply test/e2e/apimachinery/flowcontrol.go]: ah, sorry, must've missed that. fixed.

**Comment by adtac** [comment]
rebased onto latest master (resolves merge conflict)

**Comment by MikeSpreitzer** [comment]
/test pull-kubernetes-e2e-kind-ipv6

@adtac it looks like you get to remove a lint exception

**Comment by MikeSpreitzer** [comment]
LGTM except for that lint business

**Comment by adtac** [comment]
ha, this makes no sense, I got a U1000 function unused lint error in the other PR for that function, but suddenly that error seems to have disappeared; fixed.

**Comment by adtac** [comment]
/test pull-kubernetes-e2e-kind-ipv6
(unrelated sig-network flake)

**Review by MikeSpreitzer — APPROVED** [review]
/lgtm

**Comment by liggitt** [comment]
redirecting to approvers with more context per https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apimachinery/OWNERS

/assign @lavalamp 
/unassign

File: test/e2e/apimachinery/flowcontrol.go
Comment (wojtek-t) [inline test/e2e/apimachinery/flowcontrol.go:46]: seems unused

File: test/e2e/apimachinery/flowcontrol.go
Comment (wojtek-t) [inline test/e2e/apimachinery/flowcontrol.go:352]: Returning an error from this function and letting callers to handle them as they want would be a bit cleaner.

**Review by wojtek-t — COMMENTED** [review]
just two minor comments from me - other than that it looks fine 
(though before I will familiarize myself better with it, I think @lavalamp should still take a quick look)

File: test/e2e/apimachinery/flowcontrol.go
Comment (MikeSpreitzer) [reply test/e2e/apimachinery/flowcontrol.go:352]: What is meant by "cleaner" here?  The suggested change would make the code more complex.

Comment (wojtek-t) [reply test/e2e/apimachinery/flowcontrol.go:352]: Will make it a bit longer - I don't agree it's more complex.

I actually think that if something can return an error, I want to leave the decision to the called how they handle it, rather the hiding that behavior below.
As an example - the error returned FlowSchemas.Get() - I can easily imagine a caller may want to retry that in some cases. Another example is that I may want to completely differently handle the case of error from API call and differently a timeout from waiting for reconciliation.
Making a decision for the called how they should do that is not the best one.

Comment (MikeSpreitzer) [reply test/e2e/apimachinery/flowcontrol.go:352]: I understand that kind of thinking for non-test code.  For test code, it seems like unnecessary generality to me --- we do not have an open world of code that will be calling the test code.

I am not sure what complexity measure you are thinking of.  I think it is simpler to wrap up things as they are.  Returning an error requires duplicating the error-handling at all the call sites, and in this case they all want the same handling.  It is only test code; if some day we find the test code wants different error handling in different tests, we can refactor then.

But this is a pretty small detail; I can live with the suggested change.

File: test/e2e/apimachinery/flowcontrol.go
Comment (yue9944882) [inline test/e2e/apimachinery/flowcontrol.go:270]:
```suggestion
func getPriorityLevelConcurrency(c kubernetes.Interface, priorityLevelName string) (int32, error) {
```
we dont need the whole framework reference if we're returning an error in this method, amiright?

File: test/e2e/apimachinery/flowcontrol.go
Comment (wojtek-t) [reply test/e2e/apimachinery/flowcontrol.go:352]: I think it's better to change, but it's not something I'm going to fight until death too :)

Comment (lavalamp) [reply test/e2e/apimachinery/flowcontrol.go:352]: Another way to do this is to define the checking function, and let the caller invoke via the `framework.ExpectNoError(wait.Poll(...))` line. Given the number and uniformity of the invocations, though, I think that's not necessary yet.

We can always change it later, let's leave it for now.

**Comment by lavalamp** [comment]
/approve
/hold

(if you want to fix the unreferenced const)

**Comment by adtac** [comment]
Updated based on review comments. Sorry about the sig label spam.

/hold cancel

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!

**Pull request merged**
